glide-core/src/ffi/ffi_client.rs

```diff
@@ -0,0 +1,472 @@
+                error_type: error_type(&redis_error),
+            }
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime
+        .block_on(GlideClient::new(connection_request))
+        .map_err(|err| CreateClientError {
+            message: err.to_string(),
+            error_type: RequestErrorType::Disconnect,
+        })?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+fn create_client_result_to_connection_response(result: Result<Client, CreateClientError>) -> ConnectionResponse {
+    match result {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            error_message: CString::into_raw(CString::new(err.message).unwrap()),
+            error_type: err.error_type,
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            error_message: std::ptr::null(),
+            error_type: RequestErrorType::Unspecified,
+        }
+    }
+}
+
+/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+///
+/// # Safety
+///
+/// * `config` must be a valid [`ConnectionConfig`](ConnectionConfig) pointer. See the safety documentation of [`create_client_internal`](create_client_using_config_internal).
+#[allow(rustdoc::redundant_explicit_links)]
+#[allow(rustdoc::private_intra_doc_links)]
+#[no_mangle]
+pub unsafe extern "C" fn create_client_using_config(
+    config: *const ConnectionConfig,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let result = unsafe { create_client_using_config_internal(config, success_callback, failure_callback) };
+    let response = create_client_result_to_connection_response(result);
+    Box::into_raw(Box::new(response))
+}
+
+/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+///
+/// The returned `ConnectionResponse` should be manually freed by calling `free_connection_response`, otherwise a memory leak will occur. It should be freed whether or not an error occurs.
+///
+/// # Safety
+///
+/// * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes.
+/// * `connection_request_len` must not be greater than `isize::MAX`. See the safety documentation of [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+#[no_mangle]
+pub unsafe extern "C" fn create_client_using_protobuf(
```

Comment:
this requires that the caller will have a dependency on protobuf, which we want to avoid.